### PR TITLE
close aie context cleanup in shim

### PR DESCRIPTION
**Pull request by SravanKumarAllu-xilinx**
#### Problem solved by the commit

shim is getting destroyed before Aie

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered

N/A

#### How problem was solved, alternative solutions (if any) and why they were rejected

resetting Aie in shim's destructor

#### Risks (if any) associated the changes in the commit

N/A

#### What has been tested and how, request additional testing if necessary

graph test cases including profiling with device / hwcontext

#### Documentation impact (if any)

N/A

### CODE REVIEW

File: src/runtime_src/core/edge/user/shim.cpp [hunk 1]
```diff
@@ -1993,8 +1996,8 @@ void
 shim::
 registerAieArray()
 {
-  if(aieArray == nullptr)
-      aieArray = std::make_shared<zynqaie::aie_array>(mCoreDevice);
+  if(m_aie_array == nullptr)
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp:1999]:
```suggestion
  if (!m_aie_array)
```

File: src/runtime_src/core/edge/user/shim.cpp [hunk 2]
```diff
@@ -2003,7 +2006,7 @@ bool
 shim::
 isAieRegistered()
 {
-  return (aieArray != nullptr);
+  return (m_aie_array != nullptr);
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/shim.cpp:2009]: Here the comparison can be argued to clarify the meaning of what it means to be registered, so this one I would leave.

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/edge/user/aie/aie_buffer_object.cpp [hunk 3]
```diff
@@ -8,7 +8,7 @@
 #include "core/edge/user/shim.h"
 
 namespace zynqaie {
-  aie_buffer_object::aie_buffer_object(xrt_core::device* device ,const xrt::uuid uuid, const char* buffer_name, const zynqaie::hwctx_object* hwctx)
+  aie_buffer_object::aie_buffer_object(xrt_core::device* device ,const xrt::uuid uuid, const char* buffer_name, zynqaie::hwctx_object* hwctx)
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie_buffer_object.cpp:11]:
```suggestion
  aie_buffer_object::aie_buffer_object(xrt_core::device* device, const xrt::uuid uuid, const char* buffer_name, zynqaie::hwctx_object* hwctx)
```

File: src/runtime_src/core/edge/user/aie/aie_buffer_object.h [hunk 4]
```diff
@@ -23,7 +23,7 @@ namespace zynqaie {
     std::shared_ptr<aie_array> m_aie_array;
 
   public:
-    aie_buffer_object(xrt_core::device* device , const xrt::uuid uuid, const char* name, const zynqaie::hwctx_object* hwctx=nullptr);
+    aie_buffer_object(xrt_core::device* device , const xrt::uuid uuid, const char* name, zynqaie::hwctx_object* hwctx=nullptr);
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie_buffer_object.h:26]:
```suggestion
    aie_buffer_object(xrt_core::device* device, const xrt::uuid uuid, const char* name, zynqaie::hwctx_object* hwctx=nullptr);
```

File: src/runtime_src/core/edge/user/device_linux.cpp [hunk 5]
```diff
@@ -1212,24 +1212,45 @@ void
 device_linux::
 open_aie_context(xrt::aie::access_mode am)
 {
- if (auto ret = xclAIEOpenContext(get_device_handle(), am))
-   throw error(ret, "fail to open aie context");
+  auto drv = ZYNQ::shim::handleCheck(get_device_handle());
+
+  if (int ret; ret = drv->openAIEContext(am))
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/device_linux.cpp:1217]:
```suggestion
  if (int ret = drv->openAIEContext(am))
```

File: src/runtime_src/core/edge/user/device_linux.cpp [hunk 6]
```diff
@@ -1212,24 +1212,45 @@ void
...
 device_linux::
 reset_aie()
 {
-  if (auto ret = xclResetAIEArray(get_device_handle()))
-    throw system_error(ret, "fail to reset aie");
+  auto drv = ZYNQ::shim::handleCheck(get_device_handle());
+
+  if (!drv->isAieRegistered())
+    throw xrt_core::error(-EINVAL, "No AIE presented");
+  auto aie_array = drv->get_aie_array_shared();
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/device_linux.cpp:1231]:
```suggestion
  if (!drv->isAieRegistered())
    throw xrt_core::error(-EINVAL, "No AIE presented");
    
  auto aie_array = drv->get_aie_array_shared();
```

**Review by stsoe — COMMENTED** [review]
Looks good, minor tweaks.

**Review by chvamshi-xilinx — APPROVED** [review]

**Pull request merged**